src/clj/puppetlabs/puppetserver/jruby_request.clj

```diff
@@ -45,6 +46,24 @@
      jruby-puppet jruby-service {:request (dissoc request :ssl-client-cert)}
      (handler (assoc request :jruby-instance jruby-puppet)))))
 
+(defn wrap-with-concurrent-request-limit
+  "Middleware fn that returns a 503 'Service Unavailable' response if the
+  number of outstanding requests for a JRuby instance exceed the given limit.
+  If splay-limit is set, the response will include a Retry-After header set
+  to a random fraction of the limit."
+  [handler metrics-service request-limit splay-limit]
+  ;; TODO: Log warnings when limit is hit. Add i18n to user-facing strings.
+  (fn [request]
+    (let [metrics-map (jruby-metrics/get-metrics metrics-service)
+          requested-count (-> metrics-map :requested-count .getCount)]
+      (if (> requested-count request-limit)
```

Comment:
Hmm, not sure that `requested-count` is what you'd want to use for this.  I think this is an aggregate counter of all requests that have been made since service startup, not a count of the requests currently in flight.  Maybe getting the vector of `requested-instances` and doing a count of those would get you what you're looking for here, though.